book/src/dev/rfcs/0006-stolon.md

```diff
@@ -0,0 +1,446 @@
+latter goal the best we can do is to make inferences based on the current state
+of the `zcashd` source code.  Currently, `zcashd` only uses the handshake
+timestamp at the end of parsing the version message:
+```
+pfrom->nTimeOffset = timeWarning.AddTimeData(pfrom->addr, nTime, GetTime());
+```
+The `AddTimeData` method defined in `src/timedata.cpp` is a no-op as long as
+the difference between the specified timestamp and `zcashd`'s local time is less
+than a `TIMEDATA_WARNING_THRESHOLD` set to 10 minutes.  Otherwise, the data is fed
+into a process that attempts to detect when the local time is wrong.
+Truncating to the nearest 5 minutes (half of `TIMEDATA_WARNING_THRESHOLD`)
+attempts to try to stay within this interval, so as to not confuse `zcashd`.
+
+In general, truncation may not be sufficient to prevent an observer from
+obtaining distinguishing information, because of aliasing between the truncated
+data and the original signal.  For instance, continuously broadcasting the
+node's timestamp truncated to a 5-minute interval would not hide the node's
+clock skew, which would be revealed at the moment the node switched to the next
+truncation.  However, this is unlikely to be a problem in practice, because the
+number of samples an observer can obtain from a single node is likely to be
+small. And, unlike other approaches like adding random noise, truncation is
+simple, reliable, and deterministic, making it easy to test.
+
+## Deployment considerations and non-Tor relay fallback
+
+Unfortunately, Tor is difficult to use as a library and its ease of deployment
+varies by platform. We therefore need to consider how to support Tor
+integrations and to provide a fallback mechanism when Tor is not available.
+
+To handle fallback, a Zebra node without Tor access can perform the relay
+phase over an existing peer connection, rather than over an ephemeral Tor
+connection. This is much worse for privacy, but is similar to the design of
+Dandelion-lite. Ideally, Zebra+Tor is the common case for Zcash nodes, and
+Zebra nodes with Tor support can improve network privacy for Zebra without
+Tor and Zcashd nodes.
+
+We should consider how we can encourage Tor deployments for Zebra. One option
+would be to recommend Docker as the primary install mechanism, and bundle a Tor
+relay image with the Zebra image. Another would be to attempt to package Zebra
+with a dependency on Tor.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This design is more complicated than simply immediately broadcasting
+  transactions to all peers.
+
+- This design requires integration with external software (Tor) to be
+  effective, although the fallback method probably provides better privacy than
+  immediate broadcast.
+
+- Zcashd keeps IP-based peer reputation scores and may reject connections from
+  Tor exits if too many invalid transactions are sent from one exit.  However,
+  only one relay-to-Zcashd has to succeed for Zcashd to broadcast a transaction
```

Comment:
Could this allow an adversary to force a connection to eventually go through an exit node that they control?

Reply:
No, because circuit selection is performed by the local Tor instance.

Reply:
What if an attacker uses a zcashd node to poison the reputations of almost all Tor exits, except for the exits node they control?

There are approximately 600 exits that support the Zcash mainnet ports, and 300 zcashd mainnet nodes, so this attack would require on the order of 180,000 connections. (Depending on the number of connections required to poison the reputation of an IP address.)

(Some exits and zcashd nodes also have IPv6 addresses, and Tor clients allow remote IPv6 connections by default. But I don't think that significantly changes the number of connections.)

Once the reputations of most exits are poisoned, only connections via the attacker's exits will succeed.

One mitigation would be to limit the fanout, as Stolon already does - repeatedly trying a large number of exits increases the security risk here.

Reply:
Because Stolon has a limited fanout, it fails closed rather than open, so there is no privacy threat.  Zebra doesn't have this problem (as noted in the next few lines not included in the snippet above) and Zcashd's denial-of-service "defenses" can be fixed if this becomes a problem.